snapshot.go

```diff
@@ -0,0 +1,383 @@
+	} else if hdr.Name != manifestName {
+		return fmt.Errorf("invalid snapshot header: expected manifest")
+	}
+
+	// Materialize snapshot.
+	var snapshot Snapshot
+	if err := json.NewDecoder(sr.tr).Decode(&snapshot); err != nil {
+		return fmt.Errorf("decode manifest: %s", err)
+	}
+	sr.snapshot = &snapshot
+
+	return nil
+}
+
+// Next returns the next file in the snapshot.
+func (sr *SnapshotReader) Next() (SnapshotFile, error) {
+	// Read snapshot if it hasn't been read yet.
+	if err := sr.readSnapshot(); err != nil {
+		return SnapshotFile{}, err
+	}
+
+	// Read next header.
+	hdr, err := sr.tr.Next()
+	if err != nil {
+		return SnapshotFile{}, err
+	}
+
+	// Match header to file in snapshot.
+	for i := range sr.snapshot.Files {
+		if sr.snapshot.Files[i].Name == hdr.Name {
+			return sr.snapshot.Files[i], nil
+		}
+	}
+
+	// Return error if file is not in the snapshot.
+	return SnapshotFile{}, fmt.Errorf("snapshot entry not found in manifest: %s", hdr.Name)
+}
+
+// Read reads the current entry in the snapshot.
+func (sr *SnapshotReader) Read(b []byte) (n int, err error) {
+	// Read snapshot if it hasn't been read yet.
+	if err := sr.readSnapshot(); err != nil {
+		return 0, err
+	}
+
+	// Pass read through to the tar reader.
+	return sr.tr.Read(b)
+}
+
+// SnapshotWriter writes a snapshot and the underlying files to disk as a tar archive.
+type SnapshotWriter struct {
+	// The snapshot to write from.
+	// Removing files from the snapshot after creation will cause those files to be ignored.
+	Snapshot *Snapshot
+
+	// Writers for each file by filename.
+	// Writers will be closed as they're processed and will close by the end of WriteTo().
+	FileWriters map[string]SnapshotFileWriter
+}
+
+// NewSnapshotWriter returns a new instance of SnapshotWriter.
+func NewSnapshotWriter() *SnapshotWriter {
+	return &SnapshotWriter{
+		Snapshot:    &Snapshot{},
+		FileWriters: make(map[string]SnapshotFileWriter),
+	}
+}
+
+// Close closes all file writers on the snapshot.
+func (sw *SnapshotWriter) Close() error {
+	for _, fw := range sw.FileWriters {
+		_ = fw.Close()
+	}
+	return nil
+}
+
+// closeUnusedWriters closes all file writers not on the snapshot.
+// This allows transactions on these files to be short lived.
+func (sw *SnapshotWriter) closeUnusedWriters() {
+loop:
+	for name, fw := range sw.FileWriters {
+		// Find writer in snapshot.
+		for _, f := range sw.Snapshot.Files {
+			if f.Name == name {
+				continue loop
+			}
+		}
+
+		// If not found then close it.
+		_ = fw.Close()
+	}
+}
+
+// WriteTo writes the snapshot to the writer.
+// File writers are closed as they are written.
+// This function will always return n == 0.
+func (sw *SnapshotWriter) WriteTo(w io.Writer) (n int64, err error) {
+	// Close any file writers that aren't required.
+	sw.closeUnusedWriters()
+
+	// Begin writing a tar file to the output.
+	tw := tar.NewWriter(w)
+	defer tw.Close()
+
+	// Write manifest file.
+	if err := sw.writeManifestTo(tw); err != nil {
```

Comment:
I almost feel like this should be written last. Would if make checking for bad snapshots easier if it was?